### PR TITLE
Cache accepting multiple input files

### PR DESCRIPTION
**Pull request by ramarag**
@livarcocc @eerhardt @gkhanna79 PTAL

### CODE REVIEW

**Comment by ramarag** [comment]
fixes https://github.com/dotnet/cli/issues/5463

**Comment by eerhardt** [comment]
@ramarag - looks like there are test failures.  Can you fix them?

File: src/dotnet/commands/dotnet-cache/Program.cs [hunk 1]
```diff
@@ -70,13 +70,20 @@ public static CacheCommand FromArgs(string[] args, string msbuildPath = null)
...
+                if (!projectArguments.HasValue())
                 {
                     throw new InvalidOperationException(LocalizableStrings.SpecifyEntries);
                 }
 
                 msbuildArgs.Add("/t:ComposeCache");
-                msbuildArgs.Add(projectArgument.Value());
+                msbuildArgs.Add(projectArguments.Values[0]);
+
+                projectArguments.Values.RemoveAt(0);
```
Comment (eerhardt) [inline src/dotnet/commands/dotnet-cache/Program.cs]: It's best to not mutate the `projectArguments` variable.  That way if you ever need to inspect them later, you get the full list. Instead you can do something like this:

```
msbuildArgs.Add($"/p:AdditionalProjects={string.Join("%3B", projectArguments.Skip(1).Values)}");
```

Note the `Skip(1)` in there.

File: src/dotnet/commands/dotnet-cache/Program.cs [hunk 2]
```diff
@@ -106,12 +113,12 @@ public static CacheCommand FromArgs(string[] args, string msbuildPath = null)
 
                 if (skipOptimizationOption.HasValue())
                 {
-                    msbuildArgs.Add($"/p:SkipOptimization={skipOptimizationOption.HasValue()}");
+                    msbuildArgs.Add($"/p:SkipOptimization={skipOptimizationOption.HasValue().ToString().ToLower()}");
```
Comment (eerhardt) [inline src/dotnet/commands/dotnet-cache/Program.cs]: Why not something like this:

```
                if (noBuild.HasValue())
                {
                    msbuildArgs.Add($"/p:NoBuild=true");
                }

                if (includeSymbols.HasValue())
                {
                    msbuildArgs.Add($"/p:IncludeSymbols=true");
                }

                if (includeSource.HasValue())
                {
                    msbuildArgs.Add($"/p:IncludeSource=true");
                }
```

It seems more straight-forward.

File: test/Microsoft.DotNet.Tools.Tests.Utilities/Commands/CacheCommand.cs [hunk 3]
```diff
@@ -83,7 +85,7 @@ private string BuildArgs()
                 FrameworkVersionOption);
         }
 
-        private string ProfileProjectOption => string.IsNullOrEmpty(_profileProject) ? "" : $"--entries {_profileProject}";
+        private string ProfileProjectOption => _profileProject.Count == 0  ? "" :  string.Join(" ", _profileProject) ;
```
Comment (eerhardt) [inline test/Microsoft.DotNet.Tools.Tests.Utilities/Commands/CacheCommand.cs]: Won't string.Join handle the empty case as well?  I don't think you need the ternary conditional operator here.

Comment (eerhardt) [inline test/Microsoft.DotNet.Tools.Tests.Utilities/Commands/CacheCommand.cs]: Also, each profile project needs to have `--entries` on it.  That's how multi-value options work.

---
In reply to: [104684104](https://github.com/dotnet/cli/pull/5942#discussion_r104684104) [](ancestors = 104684104)

Comment (eerhardt) [inline test/Microsoft.DotNet.Tools.Tests.Utilities/Commands/CacheCommand.cs]: These comments are repeated for the PublishCommand.cs file.

---
In reply to: [104685061](https://github.com/dotnet/cli/pull/5942#discussion_r104685061) [](ancestors = 104685061,104684104)

Comment (ramarag) [inline test/Microsoft.DotNet.Tools.Tests.Utilities/Commands/CacheCommand.cs]: see the implementation of public CacheCommand WithEntries(string profileProject)

Comment (ramarag) [inline test/Microsoft.DotNet.Tools.Tests.Utilities/Commands/CacheCommand.cs]: ditto

**Comment by ramarag** [comment]
@dotnet-bot test Windows_NT x64 Release Build please

File: src/dotnet/commands/dotnet-cache/Program.cs (hunk 1 above)
Comment (ramarag) [inline src/dotnet/commands/dotnet-cache/Program.cs]: done

File: src/dotnet/commands/dotnet-cache/Program.cs (hunk 2 above)
Comment (ramarag) [inline src/dotnet/commands/dotnet-cache/Program.cs]: done

**Comment by ramarag** [comment]
@dotnet-bot test Windows_NT x64 Release Build please

**Comment by ramarag** [comment]
@dotnet-bot test CentOS7.1 x64 Debug Build

**Comment by ramarag** [comment]
@eerhardt if you have no further concerns can you go ahead and merge it ?
@piotrpMSFT still has not given me merge permissions

File: src/dotnet/commands/dotnet-cache/Program.cs [hunk 4]
```diff
@@ -70,13 +71,19 @@ public static CacheCommand FromArgs(string[] args, string msbuildPath = null)
...
-                if (string.IsNullOrEmpty(projectArgument.Value()))
+                if (!projectArguments.HasValue())
                 {
                     throw new InvalidOperationException(LocalizableStrings.SpecifyEntries);
                 }
 
                 msbuildArgs.Add("/t:ComposeCache");
-                msbuildArgs.Add(projectArgument.Value());
+                msbuildArgs.Add(projectArguments.Values[0]);
+                var _additionalProjectsargs = projectArguments.Values.Skip(1);
```
Comment (eerhardt) [inline src/dotnet/commands/dotnet-cache/Program.cs]: (nit) the [https://github.com/dotnet/corefx/blob/master/Documentation/coding-guidelines/coding-style.md](https://github.com/dotnet/corefx/blob/master/Documentation/coding-guidelines/coding-style.md) say to use `_` prefixes for fields, not local variables.

Comment (eerhardt) [inline src/dotnet/commands/dotnet-cache/Program.cs]: Can you fix this the next time you are in here?

**Pull request merged**